src/HttpSession.cpp

```diff
@@ -3,3 +3,106 @@
 // found in the LICENSE file.
 
 #include "HttpSession.h"
+
+typedef boost::asio::buffers_iterator<boost::asio::streambuf::const_buffers_type> iterator;
+std::pair<iterator, bool> match_end(iterator begin, iterator end) {
+	int index = 0;
+	const std::string match = "\r\n\r\n";
+	iterator i = begin;
+	while (i != end) {
+		if (*i == match[index])
+			index++;
+		else
+			index = 0;
+		if (index == 4)
+			return std::make_pair(i, true);
+		i++;
+	}
+	return std::make_pair(i, false);
+}
+
+void HttpSession::receiveMessage() {
+	buffer_.consume(buffer_.size());
```

Comment:
Same here, consider emptying the buffer at time of use, and assert() if it's empty here.

Reply:
Please check for concurrent run of receiveMessage().